File: protobuf/protobuf3/CSharp/SymbolTable.cs
using System;
using System.Collections.Generic;
using System.Text;

public class SymbolTable {
	private Stack<Symbol> scopeStack = new Stack<Symbol>();

    // Note: predeclared identifiers here.

    public SymbolTable() {
        var globalScope = new Symbol() { Name = "global", Classification = TypeClassification.Package_ };
        scopeStack.Push(globalScope);
    }

    public void EnterScope(Symbol newScope) {
        var current = scopeStack.Peek();
        if (newScope == current) return;
        scopeStack.Push(newScope);
    }

    public void ExitScope() {
        var current = scopeStack.Peek();
        scopeStack.Pop();
        if (scopeStack.Count == 0)
            throw new Exception();
    }

    public Symbol CurrentScope()
    {
        if (scopeStack.Count == 0) return null;
        var current_scope = scopeStack.Peek();
        return current_scope;
    }
    
    public bool Define(Symbol symbol) {
        var currentScope = CurrentScope();
        return this.DefineInScope(currentScope, symbol);
    }

    public bool DefineInScope(Symbol currentScope, Symbol symbol) {
        if (currentScope.Members.ContainsKey(symbol.Name)) {
            return false; // Symbol already defined in the current scope
        }
        symbol.Parent = currentScope;
        currentScope.Members[symbol.Name] = symbol;
        return true;
    }

    public Symbol Resolve(string name, Symbol start_scope = null)
    {
        if (start_scope == null)
        {
            foreach (Symbol scope in scopeStack)
            {
                if (scope.Members.TryGetValue(name, out Symbol symbol))
                {
                    return symbol;
                }
            }

            return null; // Symbol not found
        }
        else
        {
            if (start_scope.Members.TryGetValue(name, out Symbol symbol))
            {
                return symbol;
            }

            return null; // Symbol not found
        }
    }

    public override string ToString()
    {
        StringBuilder sb = new StringBuilder();
        foreach (var scope in scopeStack)
        {
            foreach (var member in scope.Members)
            {
                sb.AppendLine(member.ToString());
            }
        }
        return sb.ToString();
    }
}

File: protobuf/protobuf3/TypeScript/SymbolTable.ts
import { Symbol } from './Symbol.js';
import { TypeClassification } from './TypeClassification.js';

export class SymbolTable {
    private scopeStack: Symbol[] = [];

    constructor() {
        const globalScope = new Symbol("global", TypeClassification.Package_);
        this.scopeStack.push(globalScope);
    }

    enterScope(newScope: Symbol): void {
        const current = this.peek();
        if (newScope === current) return;
        this.scopeStack.push(newScope);
    }

    exitScope(): void {
        const current = this.peek();
        this.scopeStack.pop();
        if (this.scopeStack.length === 0) {
            throw new Error("Scope stack is empty");
        }
    }

    currentScope(): Symbol | null {
        if (this.scopeStack.length === 0) return null;
        return this.peek();
    }

    define(symbol: Symbol): boolean {
        const currentScope = this.currentScope();
        return this.defineInScope(currentScope, symbol);
    }

    defineInScope(currentScope: Symbol | null, symbol: Symbol): boolean {
        if (!currentScope) return false;
        if (currentScope.members.has(symbol.name)) {
            return false; // Symbol already defined in current scope
        }
        symbol.parent = currentScope;
        currentScope.members.set(symbol.name, symbol);
        return true;
    }

    resolve(name: string, startScope?: Symbol): Symbol | null {
        if (!startScope) {
            for (let i = this.scopeStack.length - 1; i >= 0; i--) {
                const scope = this.scopeStack[i];
                const symbol = scope.members.get(name);
                if (symbol) return symbol;
            }
            return null;
        } else {
            const symbol = startScope.members.get(name);
            return symbol ?? null;
        }
    }

    toString(): string {
        let sb: string[] = [];
        for (const scope of this.scopeStack) {
            for (const [_, member] of scope.members) {
                sb.push(member.toString());
            }
        }
        return sb.join("\n");
    }

    private peek(): Symbol {
        return this.scopeStack[this.scopeStack.length - 1];
    }
}
